app/controllers/documents_controller.rb

```diff
@@ -16,15 +16,44 @@ def show
   def update
     document = Document.find(params[:id])
     document.update_attributes(document_update_params(document))
+    save_base_path!(document)
     DocumentPublishingService.new.publish_draft(document)
     redirect_to document, notice: "Preview creation successful"
   rescue GdsApi::HTTPErrorResponse, SocketError => e
     Rails.logger.error(e)
     redirect_to document, alert: "Error creating preview"
   end
 
+  def generate_path
+    document = Document.find(params[:id])
+    proposed_title = params[:title]
+    base_path = generate_base_path(document, proposed_title)
+    if path_in_publishing_api?(base_path)
+      render json: { base_path: base_path, available: false }
+    else
+      render json: { base_path: base_path, available: true }
+    end
+  end
+
 private
 
+  def save_base_path!(document)
+    base_path = generate_base_path(document, document.title)
+    # TODO: Generate a base_path which isn't in use
+    return false if path_in_publishing_api?(base_path)
+    document.base_path = base_path
+    document.save!
+  end
+
+  def generate_base_path(document, title)
```

Comment:
Perhaps this method and the things in `DocumentPublishingService` can go into a separate class? Something like `PathGenerator.new(document).path` that returns a unique path?

Reply:
Makes sense - added here https://github.com/alphagov/content-publisher/pull/70/commits/016ba76ca8127115f008d8454b4b6a72ae8cda65